app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,517 @@
+            }
+            break;
+          case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:
+            // Location settings are not satisfied. However, we have no way to fix the
+            // settings so we won't show the dialog.
+            break;
+          default:
+            break;
+        }
+      }
+    });
+  }
+
+  //Advice user to turn on hotspot manually for API<26
+  void startHotspotDialog() {
+    AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+    builder.setPositiveButton(getString(R.string.go_to_settings_label), (dialog, id) -> {
+      setupWifiSettingsIntent();
+    });
+
+    builder.setNeutralButton(getString(R.string.hotspot_dialog_neutral_button), (dialog, id) -> {
+      //TO DO: START SERVER WITHIN THE SERVICE.
+      //Adding a handler because sometimes hotspot can take time to turn on.
+      //TO DO: Add a progress dialog instead of handler
+      ProgressDialog progressDialog =
+          ProgressDialog.show(this, getString(R.string.progress_dialog_starting_server), "",
+              true);
+      progressDialog.show();
+      final Handler handler = new Handler();
+      handler.postDelayed(new Runnable() {
+        @Override
+        public void run() {
+          progressDialog.dismiss();
+          startService(ACTION_START_SERVER);
+        }
+      }, 7000);
+    });
+
+    builder.setTitle(getString(R.string.hotspot_dialog_title));
+    builder.setMessage(
+        getString(R.string.hotspot_dialog_message)
+    );
+    AlertDialog dialog = builder.create();
+    dialog.show();
+  }
+
+  void startService(String ACTION) {
+    serviceIntent.setAction(ACTION);
+    this.startService(serviceIntent);
+  }
+
+  void mobileDataDialog() {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+      AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+      builder.setPositiveButton(this.getString(R.string.yes),
+          (dialog, id) -> openDataUsageActivity());
+      builder.setNegativeButton((android.R.string.no), (dialog, id) -> {
+        startHotspotDialog();
+      });
+      builder.setTitle(this.getString(R.string.mobile_data_enabled));
+      builder.setMessage(
+          this.getString(R.string.mobile_data_message) + "\n" + this.getString(
+              R.string.mobile_data_message_confirmation)
+      );
+      builder.setCancelable(false);
+      AlertDialog dialog = builder.create();
+      dialog.show();
+    }
+  }
+
+  @Override public void serverStarted(String ip) {
+    this.ip = ip;
+    serverTextView.setText(getString(R.string.server_started_message) + " " + this.ip);
+    startServerButton.setText(getString(R.string.stop_server_label));
+    startServerButton.setBackgroundColor(getResources().getColor(R.color.stopServer));
+    isServerStarted = true;
+  }
+
+  @Override public void serverStopped() {
+    serverTextView.setText(getString(R.string.server_textview_default_message));
```

Comment:
you don't need to call getString in nearly all the palces you call it, setText takes a resId too

Reply:
Tried this eca5f2a035aa560b3702e1565ff4e5ef6caeb64d but it's printing the id integers. Had to revert it back.

Reply:
There is just no way that can be happening. https://developer.android.com/reference/android/widget/TextView.html#setText(int)
```
serverTextView.setText(R.string.server_started_message + " " + savedInstanceState.getString(
             IP_STATE_KEY));
```
what you did here was get an int `R.string.server_started_message` and add it a string preforming string concatenation.

Reply:
Fixed. 50e072b